Explain the use of `prefer-diff-as-parent` (#2286)
neo4j/docs-operations: modules/ROOT/pages/backup-restore/online-backup.adoc

----
======
=====

[role=label--new-2025.04]
[[diff-backup-as-parent]]
=== Perform a differential backup using the `--prefer-diff-as-parent` option

> When creating a differential backup, the parent is always the most recent (non-empty) backup in the directory.
> In some cases the directory may have two equally recent backups -- a full backup and a differential one.
> By default, the full backup is chosen as the parent.
> But you can use the differential backup as the parent by setting the option `--prefer-diff-as-parent` to `True`.

Reviewer: ```suggestion
When creating a differential backup, the parent is usually the most recent (non-empty) backup in the directory.
In some cases you may prefer to use the latest differential backup as the parent, even if there is a more recent full backup.
You can use the latest differential backup as the parent by setting the option `--prefer-diff-as-parent` to `True`.
This can be used to ensure you have differential backups for all transactions, which would allow you to restore to any point in time.
Otherwise the transactions between a full backup and the previous differential backup will not be backed up as individual transactions.
```

Author: Thank you, @nick-giles-neo, for a super quick response 😄

----
======
=====

[role=label--new-2025.04]
[[diff-backup-as-parent]]
=== Perform a differential backup using the `--prefer-diff-as-parent` option

When taking a differential backup with the `--type=DIFF` option, the parent, by default, is the *most recent non-empty* backup in the directory.

> In some cases, you may prefer to use the *latest differential* backup as the parent, even if there is a more recent full backup.

Reviewer 2: What are these cases? If we don't specify, then maybe this part doesn't add much value. So, either we say something like:
```suggestion
However, in some cases, such as ...., you may prefer to use the *latest differential* backup as the parent, even if there is a more recent full backup.

```
or directly say how to use it like:

```
However, you can also use the *latest differential* backup as the parent, even when there is a more recent full backup, by setting the option `--prefer-diff-as-parent` to `true`.

> You can do this by setting the option `--prefer-diff-as-parent` to `True`.

Reviewer 2: We try not to use "You can do this".

> This can be used to ensure you have differential backups for all transactions, which would allow you to restore to any point in time.

Reviewer 2: I would actually move this to the top of the section and start with it. Something like:

```
You can use the `--prefer-diff-as-parent` option to ensure you have differential backups for all transactions and can restore to any point in time. 
It uses the latest differential backup as the parent, even if there is a more recent full backup. 
By default, the differential backup (`--type=DIFF`) uses the most recent non-empty backup in the directory as the parent. However, the transactions between a full backup and the previous differential backup cannot be backed up as individual transactions.

To use the --prefer-diff-as-parent option, set it to `true`.
```

Reviewer 2: The above is my suggestion for an intro.

Reviewer 2: It's still not perfect, but I think it should be in that line of thought.

Author: @renetapopova, thanks for your suggestions! I'll use it but change a little bit. I still think it's better to tell about the default behaviour first and then explain the use of the new option.

Reviewer 2: It looks great!

Otherwise, the transactions between a full backup and the previous differential backup will not be backed up as individual transactions.

> The examples below cover different scenarios for using the `--prefer-diff-as-parent` option.

Reviewer 2: ```suggestion
The following examples cover different scenarios for using the `--prefer-diff-as-parent` option.
```

[.tabbed-example]
=====
[role=include-with-Chain-with-full-and-differential-backups]
======

> Let's assume that every hour you write 10 transactions to the `neo4j` database, except from 12:30-13:30, when you do not write any transaction.

Reviewer 2: ```suggestion
Let's assume that you write 10 transactions to the `neo4j` database every hour, except from 12:30 to 13:30, when you do not write any transactions.
```

There is a backup job that takes a backup every hour and a full backup every four hours.

> We refer as an _empty_ backup a backup that has no transactions, meaning that both the lower transaction ID and the upper transaction ID are zero.

Reviewer 2: ```suggestion
An empty backup has no transactions, meaning that both the lower transaction ID and the upper transaction ID are zero.
```

Imagine you have the following backup chain:

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 1
| 10

| 11:30
| backup2
| DIFF
| 11
| 20

| 12:30
| backup3
| DIFF
| 21
| 30

| 13:30
| backup4
| DIFF
| 0
| 0

| 14:30
| backup5
| FULL
| 1
| 40

|===

At 15:30, you execute the following backup command:

[source,shell]
----
neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF neo4j
----

The result would be:

[cols="h,e,m,h,h"]
|===
| 14:30
| backup6
| DIFF
| 41
| 50
|===

The result means you have chosen `backup5` as the parent for your differential `backup6` since the `backup5` is the *latest non-empty* backup.

However, if you execute the following command with the `--prefer-diff-as-parent` option:

[source,shell]
----
neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF --prefer-diff-as-parent neo4j
----

The result would be:

[cols="h,e,m,h,h"]
|===

> | 14:30

Author: ```suggestion
| 15:30
```

======
[role=include-with-Chain-with-only-full-backups]
======

Let's assume that you write 10 transactions to the `neo4j` database every hour and trigger an hourly full backup.

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 1
| 10

| 11:30
| backup2
| FULL
| 11
| 20
|===

In this case, there is no differential backup.
Therefore, the `--prefer-diff-as-parent` option has no effect and the behaviour is the same as the default one.

[source,shell]
----
neo4j-admin database backup \
--from=<address:port> --to-path=<targetPath> \
--type=DIFF --prefer-diff-as-parent \
neo4j
----

The result would be (with or without the `--prefer-diff-as-parent` option):
[cols="h,e,m,h,h"]
|===
| 12:30
| backup3
| DIFF
| 21
| 30
|===

======
[role=include-with-Chain-with-only-empty-full-backups]
======

Let's assume that the database is empty and you do not write anything to it, while still taking hourly full backups.

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 0
| 0

| 11:30
| backup2
| FULL
| 0
| 0
|===

> In this case, you cannot perform a differential backup with the `--type=DIFF` option, and the below command fails anyway, whether you used the `--prefer-diff-as-parent` or not.

Reviewer 2: ```suggestion
In this case, you cannot perform a differential backup with the `--type=DIFF` option, and the below command will fail anyway, whether you use the `--prefer-diff-as-parent` or not.
```

> This occurs because we are looking for the *latest non-empty* backup, and there are only empty backups.

Reviewer 2: ```suggestion
This occurs because you are looking for the *latest non-empty* backup, and there are only empty backups.
```

[source,shell]
----
neo4j-admin database backup \
--from=<address:port> --to-path=<targetPath> \
--type=DIFF --prefer-diff-as-parent \
neo4j
----

> But if you select the `--type=AUTO` option, the command will succeed, and the result would be another empty *full* backup.

Reviewer 2: ```suggestion
But if you select the `--type=AUTO` option, the command will succeed, and the result will be another empty *full* backup.
```

----
======
=====

[role=label--new-2025.04]
[[diff-backup-as-parent]]
=== Perform a differential backup using the `--prefer-diff-as-parent` option

By default, a differential backup (`--type=DIFF`) uses the *most recent non-empty* backup -- whether full or differential -- in the directory as its parent. 

> The `--prefer-diff-as-parent` option changes this behavior and allows you to use the *latest differential* backup as its parent, even if a newer full backup exists. 

Author: ```suggestion
The `--prefer-diff-as-parent` option changes this behavior and forces the backup job to use the *latest differential* backup as its parent, even if a newer full backup exists. 
```

> This apprroach allows you to maintain a chain of differential backups for all transactions and restore to any point in time.

Reviewer: Typo
```suggestion
This approach allows you to maintain a chain of differential backups for all transactions and restore to any point in time.
```

To use the `--prefer-diff-as-parent` option, set it to `true`.

The following examples cover different scenarios for using the `--prefer-diff-as-parent` option.

[.tabbed-example]
=====
[role=include-with-Chain-with-full-and-differential-backups]
======

Let's assume that you write 10 transactions to the `neo4j` database every hour, except from 12:30 to 13:30, when you do not write any transactions.

There is a backup job that takes a backup every hour and a full backup every four hours.
An empty backup has no transactions, meaning that both the lower transaction ID and the upper transaction ID are zero.

Imagine you have the following backup chain:

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 1
| 10

| 11:30
| backup2
| DIFF
| 11
| 20

| 12:30
| backup3
| DIFF
| 21
| 30

| 13:30
| backup4
| DIFF
| 0
| 0

| 14:30
| backup5
| FULL
| 1
| 40

|===

At 15:30, you execute the following backup command:

[source,shell]
----
neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF neo4j
----

The result would be:

[cols="h,e,m,h,h"]
|===

> | 14:30

Author: ```suggestion
| 15:30
```

| DIFF
| 41
| 50
|===

The result means that the `backup5` is the parent for your differential `backup6` since the `backup5` is the *latest non-empty* backup.

However, if you execute the following command with the `--prefer-diff-as-parent` option:

[source,shell]
----

> neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF --prefer-diff-as-parent neo4j

Author: ```suggestion
neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF --prefer-diff-as-parent=true neo4j
```

Author: Should we set `prefer-diff-as-parent` to `true` here?

neo4j-admin database backup --from=<address:port> --to-path=<targetPath> --type=DIFF --prefer-diff-as-parent neo4j
----

The result would be:

[cols="h,e,m,h,h"]
|===

> | 14:30

Author: ```suggestion
| 15:30
```

| backup6
| DIFF
| 31
| 50
|===

In this case, the `backup3` is selected as the parent since it is the *latest non-empty differential* backup.

======
[role=include-with-Chain-with-only-full-backups]
======

Let's assume that you write 10 transactions to the `neo4j` database every hour and trigger an hourly full backup.

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 1
| 10

| 11:30
| backup2
| FULL
| 11
| 20
|===

In this case, there is no differential backup.
Therefore, the `--prefer-diff-as-parent` option has no effect and the behaviour is the same as the default one.

[source,shell]
----
neo4j-admin database backup \
--from=<address:port> --to-path=<targetPath> \
--type=DIFF --prefer-diff-as-parent \
neo4j
----

The result would be (with or without the `--prefer-diff-as-parent` option):
[cols="h,e,m,h,h"]
|===
| 12:30
| backup3
| DIFF
| 21
| 30
|===

======
[role=include-with-Chain-with-only-empty-full-backups]
======

> Let's assume that the database is empty and you do not write anything to it, while still taking hourly full backups.

Reviewer: I don't think this is possible, you cannot have empty full backups. Even if the database has never had anything written to it, there is at least a bootstrap transaction, possibly a few. So the full backup will be from 1 to 2 or 3.

In this case, there is no differential backup.
Therefore, the `--prefer-diff-as-parent` option has no effect and the behaviour is the same as the default one.

[source,shell]
----
neo4j-admin database backup \
--from=<address:port> --to-path=<targetPath> \
--type=DIFF --prefer-diff-as-parent \
neo4j
----

The result would be (with or without the `--prefer-diff-as-parent` option):
[cols="h,e,m,h,h"]
|===
| 12:30
| backup3
| DIFF
| 21
| 30
|===

======
[role=include-with-Chain-with-only-empty-full-backups]
======

Let's assume that the database is empty and you do not write anything to it, while still taking hourly full backups.

[cols="h,e,m,h,h"]
|===
|Timestamp | Backup name | Backup type | Lower Transaction ID | Upper Transaction ID

| 10:30
| backup1
| FULL
| 0
| 0

| 11:30
| backup2
| FULL
| 0
| 0
|===

In this case, you cannot perform a differential backup with the `--type=DIFF` option, and the below command fails anyway, whether you used the `--prefer-diff-as-parent` or not.
This occurs because we are looking for the *latest non-empty* backup, and there are only empty backups.

[source,shell]
----
neo4j-admin database backup \
--from=<address:port> --to-path=<targetPath> \

> --type=DIFF --prefer-diff-as-parent \

Author: ```suggestion
--type=DIFF --prefer-diff-as-parent=true \
```

Reviewer 3: Same as above, can be omitted the =true
